feat: LEAP-1896: Add scrollbar support to Visualizer

Gondragos opened the pull request:
Introduced a scroll filler element to improve handling of browser scroll size inconsistencies in the Visualizer. Updated relevant methods and styles to adjust the layout dynamically based on the calculated scroll size. Added safeguards to mouse events.
<img width="1152" alt="image" src="https://github.com/user-attachments/assets/633d8efc-27f3-4e7d-8fa9-fc4ecf61272d" />

This pull request introduces several enhancements and fixes to the `Visualizer` and `Waveform` components, as well as some updates to the testing helpers. The most significant changes include the addition of a scrollbar, improvements to the `Visualizer`'s scroll handling, and updates to ensure the `AudioUltra`'s readiness state.

### Enhancements to Visualizer and Waveform:
* Added `BROWSER_SCROLLBAR_WIDTH` constant for dynamic scrollbar width calculation in `Utils.ts`.
* Introduced `scrollFiller` element and `initScrollBar` method to manage scrollable area size in `Visualizer.ts`. [[1]](diffhunk://#diff-840324f862f81f5870f0146d988d96794250c39ac63f53ba1d75301e7288ec68R51) [[2]](diffhunk://#diff-840324f862f81f5870f0146d988d96794250c39ac63f53ba1d75301e7288ec68L642-R697)
* Splited `setScrollLeft` and `_setScrollLeft` methods to handle scroll position updates in `Visualizer.ts`.
* Added `redrawCursor` method to improve synced cursor rendering in `Visualizer.ts`.
* Updated `centerToCurrentTime` method to use `setScrollLeft` to keep the scrollbar up to date.

### Improvements to readiness and event handling:
* Added `checkReady` method to ensure `Waveform` readiness state in `model.js`. (There were some issues capturing the final drawing state after the audio finished loading, especially in "view all" mode and tests.)
* Updated `syncCursor` method to use `redrawCursor` for cursor updates in `Waveform.ts`. (to avoid redrawing all layers)
* Added `isDrawing` getter to `Waveform` for checking drawing state in `Waveform.ts`.

### Testing helpers updates:
* Added `waitForObjectsReady` call in `isReady` method to ensure readiness in `AudioView.ts`. [[1]](diffhunk://#diff-885d3156d22169fe1e0b3d76328450f5eafc79413bef88294f37e76a6581edf7R4) [[2]](diffhunk://#diff-885d3156d22169fe1e0b3d76328450f5eafc79413bef88294f37e76a6581edf7R40)

Gondragos committed:
feat: LEAP-1896: Add scrollbar support to Visualizer

Introduced a scroll filler element to improve handling of browser scroll size inconsistencies in the Visualizer. Updated relevant methods and styles to adjust layout dynamically based on the calculated scroll size. Added safeguards to mouse events.

hlomzik commented on web/libs/editor/src/lib/AudioUltra/Common/Utils.ts:
```diff
@@ -212,3 +212,27 @@ export const getCursorTime = (e: MouseEvent, visualizer: Visualizer, duration: n
   isTimeSimilar(a / observedDuration, b / observedDuration);
+
+/**
+ * A constant representing the width of the browser's scrollbar in pixels.
```
I'd correct this to "thickness of scrollbar's handle" to be precise and rename the constant name. Because "scroll width" is basically a client width in browser terms.

Gondragos replied:
There is CSS property `scrollbar-width` with almost the same meaning that I mean. But I still do not understand, if `BROWSER_SCROLLBAR_WIDTH` helps or not.

Gondragos committed:
fix: Rename scrollbar size constant for clarity

Renamed `BROWSER_SCROLL_SIZE` to `BROWSER_SCROLLBAR_WIDTH` to provide a clearer description of its purpose. Updated related references and comments to reflect this change, ensuring consistency and maintaining accurate documentation.

Gondragos committed:
Improve scrollbar handling for audio visualizer

Refactored scrollbar logic into a new `initScrollBar` method for better clarity and functionality. Added `scrollbarGutter: stable` for potential consistent scrollbar behavior across browsers.

Gondragos committed:
fix: adjust pixel color comparison for consistent tests

Updated pixel color comparison coordinates in audio region tests to ensure accuracy and consistency. Added a getter to check the readiness of objects in the model and updated `isReady` logic in test helpers for smoother synchronization.

Gondragos committed:
fix: ensure audio view fully renders before proceeding

Added a short wait after verifying the loading bar's absence to ensure the AudioView has completely rendered. This change addresses potential flakiness in tests caused by premature assertions.

hlomzik commented on web/libs/frontend-test/src/helpers/LSF/AudioView.ts:
```diff
@@ -36,7 +37,9 @@ export const AudioView = {
   isReady() {
+    LabelStudio.waitForObjectsReady();
     this.loadingBar.should("not.exist");
+    cy.wait(32); // wait for render
```
can we use any kind of observer to wait for actual audio rendered? 32s is too long...

Gondragos replied:
It is in milliseconds. Like approximately two cycles of rendering.

Gondragos replied:
But yeah, we can add some code into `AudioUltra` to make `isReady` work better, but for now it feels like too much, as we did not add it when it was necessary. And right now I have no free time in the boundaries of estimations for this task.

bmartel replied:
I would say adding as a comment what you just explained would be good enough for now 👍

Gondragos committed:
test: Revert changes and add comments

MihajloHoma commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/13696078702)
> Successfully merged:  6 files changed, 86 insertions(+), 21 deletions(-)

MihajloHoma committed:
Merge branch 'develop' into 'fb-LEAP-1896/audio-scroll'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13696078702

Gondragos committed:
fix: Revert isReady redefining

It looks like it was not the problem for tests.

Gondragos committed:
fix: adjust comment regarding `isReady` timing in AudioView

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/13709067876)
> Successfully merged:  create mode 100644 web/tsconfig.json

Gondragos committed:
Merge branch 'develop' into 'fb-LEAP-1896/audio-scroll'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13709067876

MihajloHoma commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/13758696043)
> Successfully merged:  11 files changed, 21 insertions(+), 17 deletions(-)

MihajloHoma committed:
Merge branch 'develop' into 'fb-LEAP-1896/audio-scroll'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13758696043

Gondragos commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/13782129545)
> Successfully merged:  create mode 100644 web/libs/ui/src/lib/enterprise-badge/enterprise-badge.tsx

Gondragos committed:
Merge branch 'develop' into 'fb-LEAP-1896/audio-scroll'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13782129545

Gondragos committed:
fix: improve cursor sync logic in waveform editor

Replaced redundant full redraw with a targeted cursor update. Added `redrawCursor` method in Visualizer for efficient cursor rendering, enhancing performance and reducing unnecessary visual updates.

Gondragos committed:
fix: remove unnecessary scrollbarGutter style

Gondragos committed:
fix: ensure readiness checks account for drawing state

Previously, `onReady` was called without considering the `isDrawing` state, which could lead to premature execution. Introduced `checkReady` to delay readiness until drawing completes, ensuring proper timing and avoiding potential issues.

Gondragos committed:
fix: correct rendering sequence in Visualizer resize handler

Reordered method calls in the Visualizer update loop to ensure proper rendering. Moved `updateScrollFiller` and `setScrollLeft` to prevent visual inconsistencies. This resolves timing issues when updating the cursor position and timeline.

Gondragos commented:
/git merge develop
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/13815671005)
> Successfully merged:  delete mode 100644 label_studio/templates/401.html

Gondragos committed:
Merge branch 'develop' into 'fb-LEAP-1896/audio-scroll'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/13815671005
